### PR TITLE
Implements RecipeSelector

### PR DESCRIPTION
**Pull request by raulverag**
With tests.

### CODE REVIEW

**Comment by raulverag** [comment]
This still needs to ensure that values in actual AiAi requests are supported, e.g. dotted strings for apk names. That will come in the next PR.

File: src/recipe-selector/recipe-selector.ts [hunk 1]
```diff
@@ -11,14 +11,54 @@
...
+    // Other recipes are silently ignored. This relies on the triggers array
+    // being always present but empty if there are no triggers.
+    // The order of the Matches in the lookup table preserves the order of the
+    // Recipes in the constructor argument, and within each recipe the order of
+    // the triggers in the triggers field of the recipe.
     constructor(readonly recipes: Recipe[]) {
+      recipes.forEach(recipe => {
+        recipe.triggers.forEach(trigger => {
+          if (this._table.find(match => match.matches(trigger))) {
+            throw new Error('Duplicate recipe trigger!');
```
Comment (mmandlis) [inline src/recipe-selector/recipe-selector.ts]: in the design doc you mentioned checking recipes for subset triggers, but matches works in only one direction, is this intentional?

File: src/recipe-selector/recipe-selector.ts [hunk 2]
```diff
@@ -11,14 +11,54 @@
...
+    }
+
     // Only recipes with “@trigger” annotations get included in the lookup table.
-    // Other recipes are silently ignored.
-    // Specific semantics of overlapping triggers are still TBD.
+    // Other recipes are silently ignored. This relies on the triggers array
+    // being always present but empty if there are no triggers.
+    // The order of the Matches in the lookup table preserves the order of the
+    // Recipes in the constructor argument, and within each recipe the order of
+    // the triggers in the triggers field of the recipe.
```
Comment (mmandlis) [inline src/recipe-selector/recipe-selector.ts]: RecipeSelector could instead take Manifest as ctor parameter and own the logic of which recipes should be used.

File: src/recipe-selector/recipe-selector.ts [hunk 3]
```diff
@@ -11,14 +11,54 @@
...
+      });
     }
-    // Returns a Recipe or null if no trigger matches the request)
-    select(trigger: [string][]) : Recipe | null {
+
+    // Returns a Recipe or null if no trigger matches the request
+    // The table is consulted in order and the first successful match is returned.
+    select(request: [string, string][]) : Recipe | null {
+      const found = this._table.find(match => match.matches(request));
+      if (found) return found.recipe;
```
Comment (mmandlis) [inline src/recipe-selector/recipe-selector.ts]: maybe log a warning, if there is more than 1?

File: src/recipe-selector/recipe-selector.ts (hunk 1 above)
Comment (raulverag) [reply src/recipe-selector/recipe-selector.ts]: That changed, as it wasn't sufficient and too restrictive. The doc has been updated. Duplicates are disallowed, and the first match wins. This allows more specific triggers to come first, e.g. "app foo, action bar", with a more general catch-all later, e.g. "app foo".

File: src/recipe-selector/recipe-selector.ts (hunk 2 above)
Comment (raulverag) [reply src/recipe-selector/recipe-selector.ts]: There actually isn't much logic. The triggers are part of the recipes, initialized to an empty array and filled if any triggers are encountered when the manifest is parsed. The parsing does already preserves both orders, and I've included a test to ensure that this remains the case. So this code just preserves order when building the lookup table, and ignores recipes with empty trigger arrays. So unless the parsing changes to do something to the order, there really isn't anything to be gained by taking the manifest on the constructor.

File: src/recipe-selector/recipe-selector.ts (hunk 3 above)
Comment (raulverag) [reply src/recipe-selector/recipe-selector.ts]: Now that the matching can be partial (i.e. the whole trigger must match, but that can be only part of the request), this is actually a feature. A very specific trigger, with several key-value pairs, can be placed before one that's more general, with fewer of the same pairs. A request matching the very specific trigger would also match the more general one, so matching more than one trigger is no longer a problem at all.

**Pull request merged**